pkg/engine/renderer/enhancer.go

```diff
@@ -70,24 +72,66 @@ func (de *DefaultEnhancer) Apply(templates map[string]string, metadata Metadata)
 				}
 			}
 
-			// This is pretty important, if we don't convert it back to the original type everything will be Unstructured.
-			// We depend on types later on in the processing e.g. when evaluating health.
-			// Additionally, as we add annotations and labels to all possible paths, this step gets rid of anything
-			// that doesn't belong to the specific object type.
+			// We convert to the typed version here and back to clean additional labels and annotations that were
+			// added. This needs to be done as otherwise the dependency calculator calculates hashes based on the added
+			// fields which are later missing if the resource is fetched from the api-server
 			err = runtime.DefaultUnstructuredConverter.FromUnstructured(objUnstructured.UnstructuredContent(), obj)
 			if err != nil {
 				return nil, fmt.Errorf("%wconverting from unstructured failed: %v", engine.ErrFatalExecution, err)
 			}
-			objs = append(objs, obj)
+			unstructMap, err = runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+			if err != nil {
+				return nil, fmt.Errorf("%wconverting to unstructured failed: %v", engine.ErrFatalExecution, err)
+			}
+			objUnstructured = &unstructured.Unstructured{Object: unstructMap}
+
+			unstructuredObjs = append(unstructuredObjs, objUnstructured)
 		}
 	}
 
+	dc := newDependencyCalculator(de.Client, unstructuredObjs)
```

Comment:
This block could be a separate method e.g. `setDependenciesHashes` (the name is very debatable)